Update Substrate to 1.8.0

nazar-pc opened the pull request:
This PR updates Substrate to polkadot-sdk version 1.8.0. 1.9.0 will be updated separately due to more breaking changes there.

There were minor breaking changes that resulted in `subspace-v4` branch in polkadot-sdk fork and `subspace-v6` in frontier fork, but nothing major. Going forward to decrease maintenance burden around frontier I plan to only upgrade to polkadot-sdk releases when frontier is update to them so I don't have to fix compatibility myself that they will not merge upstream afterwards (frontier is always lagging behind).

I had to break `enable_non_root_calls` (and probably remove eventually) due to `Sudo::key()` being removed and other relevant information to identify if call is done by root or not seemingly not present in `SignedExtension` API.

The biggest changes this time came to RPC layer, but were mostly mechanical. `TransactionExtension` is something Gav authored that will likely cause more work upgrading in the future, but doesn't impact us as much for now.

Sync changes were not too breaking for our fork, but I like the direction they are moving, @shamil-gadelshin please take a look there.

As to migrations and hooks, they recommend to reduce `on_initialize` logic to the minimum, I don't think it is a hard requirement for now, but something to keep in mind going forward.

And I removed all versions from Substrate dependencies, they changed too many of them and it was painful and a bit pointless to update everywhere.

Interesting PRs on Substrate side:
* rpc: backpressured RPC server (bump jsonrpsee 0.20): https://github.com/paritytech/polkadot-sdk/pull/1313
* Prepare syncing for parallel sync strategies: https://github.com/paritytech/polkadot-sdk/pull/3224
* rpc: bump jsonrpsee v0.22 and fix race in rpc v2 chain_head: https://github.com/paritytech/polkadot-sdk/pull/3230
* Multi-Block-Migrations, poll hook and new System callbacks: https://github.com/paritytech/polkadot-sdk/pull/1781
* Remove as frame_system::DefaultConfig from the required syntax in derive_impl: https://github.com/paritytech/polkadot-sdk/pull/3505
* FRAME: Create TransactionExtension as a replacement for SignedExtension: https://github.com/paritytech/polkadot-sdk/pull/2280
* removed pallet::getter from pallet-sudo: https://github.com/paritytech/polkadot-sdk/pull/3370

The only potentially interesting PRs on frontier side are these:
* Use previous block hash to compute tx effectiveGasPrice: https://github.com/polkadot-evm/frontier/pull/1280
* allow multi account conversion: https://github.com/polkadot-evm/frontier/pull/1295

P.S. Running this version locally already, seems to work so far.

### Code contributor checklist:
* [x] I have read, understood and followed [contributing guide](https://github.com/subspace/subspace/blob/main/CONTRIBUTING.md)

nazar-pc committed:
Update Substrate to 1.8.0

nazar-pc committed:
Tiny cleanup

shamil-gadelshin reviewed (dismissed):
> Sync changes were not too breaking for our fork, but I like the direction they are moving

I understood the changes from sync-related PR as a "preparation of common steps for different sync strategies that can be shared". If I understood it correctly then the new strategy can be composed out of existing steps and it can be convenient.

nazar-pc commented on crates/pallet-domains/src/tests.rs:
```diff
@@ -133,11 +132,10 @@ impl pallet_domains::HoldIdentifier<Test> for HoldIdentifier {
 
 impl VariantCount for HoldIdentifier {
-    const VARIANT_COUNT: u32 = mem::variant_count::<Self>() as u32;
+    const VARIANT_COUNT: u32 = 10;
```
This is wrong though, check the description of the trait. It is supposed to return the same number as number of enum variants. Substrate can't use `mem::variant_count` because they rely on stable compiler, but we can because we use nightly.

NingLin-P replied:
Yeah, this violates the usage of the trait, but the max number of hold limits is hardcoded to `HoldIdentifier::VARIANT_COUNT` in [pallet-balances](https://github.com/paritytech/polkadot-sdk/blob/216509dbaa2c2941ee75fbcc9a086deab5e2c8a6/substrate/frame/balances/src/lib.rs#L476-L485), which doesn't fit into our usage (even in test) so we need this hack.

nazar-pc replied:
This is a hack on top of the hack. If it works by accident right now, I'm fine having it temporarily with a TODO explaining why we are hardcoding it to the wrong number. And we should get rid of it before it bites us again.

NingLin-P replied:
For the reviewer, see more discussion at https://github.com/subspace/subspace/issues/2674. PR is forced push to add TODO comment.

NingLin-P committed:
Update VARIANT_COUNT for HoldIdentifier

The value is used as the max number of hold an account can create, it is set to 100
for now to workaround some issues with this limit, should revist the value later.

Signed-off-by: linning <linningde25@gmail.com>
